mozilla_aws_cli/listener.py

```diff
@@ -162,48 +194,35 @@ def handle_oidc_redirect_callback():
 
     # callback into the login.callback() function in login.py
     logger.debug("redirect_callback : request is {}".format(request.json))
-    login.get_id_token(**request.json)
-    login.validate_id_token()
-    logger.debug("id_token_dict : {}".format(login.id_token_dict))
-    if login.id_token_dict is None:
-        logger.debug('Validation of token failed : {}'.format(login.token))
-        # TODO : What should we do in this case? How should the UI handle this?
+    if not login.get_id_token(**request.json):
+        return jsonify({
+            "result": "error",
+            "status_code": 400,
+        })
+    if login.validate_id_token() is None:
         return jsonify({
             "result": "id_token_validation_failed",
             "status_code": 400,
         })
+    if login.role_arn is None:
+        if login.batch:
+            login.exit('No role_arn provided. Exiting due to batch mode.')
+        else:
+            login.state = "role_picker"
+        return jsonify({
+            "result": login.state,
+            "status_code": 200,
+        })
 
-    login.get_role_map()
-    try:
-        # TODO: Consider whether this needs to loop forever
-        while login.credentials is None:
-            login.exchange_token_for_credentials()
-    except STSWarning as e:
-        if e.args[1] == 'ExpiredTokenException':
-            logger.debug('AWS says that the ID token is expired : {}'.format(e[2]))
-            login.token = None
-            url_parameters = {
-                "scope": login.oidc_scope,
-                "response_type": "code",
-                "redirect_uri": login.redirect_uri,
-                "client_id": login.client_id,
-                "code_challenge": login.code_challenge,
-                "code_challenge_method": "S256",
-                "state": login.oidc_state,
-            }
-            url = "{}?{}".format(login.authorization_endpoint,
-                                 urlencode(url_parameters))
-            logger.debug('Setting state to restart_auth and idpUrl to {}'.format(url))
-            login.state = "restart_auth"
-            login.web_state["idpUrl"] = url
-            return jsonify({
-                "result": "restart_auth",
-                "status_code": 200,
-            })
-    except requests.exceptions.ConnectionError as e:
-        login.exit("Unable to contact AWS : {}".format(e))
-
-    login.print_output()
+    if not login.get_role_map():
```

Comment:
Not storing the role_map is probably okay, but would it ever mean calling out to the end point twice? Is it possible for it to expire between calls, or should it always grab it from cache?

Reply:
`get_role_map` is called in 4 places
* The user passes a role_arn on the command line
  * We have a valid cached ID token
    * We have to build the awscli profile name we need the alias from the role map
    * We attempted to get credentials for a given role, got AccessDenied, so we're fetching the role map again (with cache=False) because maybe ours is out of date
  * We don't have a valid cached ID token
    * We have to build the awscli profile name we need the alias from the role map
* The user does not pass a role_arn on the command line
  * The user spawns the role picker and so the role map is fetched to show the available roles

So each of these conditions we do actually want to call it and it won't result in multiple calls to the endpoint in a single run